clang/lib/Sema/SemaOpenMP.cpp

```diff
@@ -5965,6 +5967,269 @@ static bool teamsLoopCanBeParallelFor(Stmt *AStmt, Sema &SemaRef) {
+      else
+        NewDecl->setContextParam(I, CDecl->getContextParam());
+    }
+
+    // Create a New Captured Stmt containing the New Captured Decl
+    SmallVector<CapturedStmt::Capture, 4> Captures;
+    SmallVector<Expr *, 4> CaptureInits;
+    for (const CapturedStmt::Capture &Capture : AssocStmt->captures())
+      Captures.push_back(Capture);
+    for (Expr *CaptureInit : AssocStmt->capture_inits())
+      CaptureInits.push_back(CaptureInit);
+    auto *NewStmt = CapturedStmt::Create(
+        Context, AssocStmt->getCapturedStmt(),
+        AssocStmt->getCapturedRegionKind(), Captures, CaptureInits, NewDecl,
+        const_cast<RecordDecl *>(AssocStmt->getCapturedRecordDecl()));
+
+    ResultAssocStmt = NewStmt;
+  }
+  return ResultAssocStmt;
+}
+
+/// replaceWithNewTraitsOrDirectCall() is for transforming the call traits.
+/// Call traits associated with a function call are removed and replaced with
+/// a direct call. For clause "nocontext" only, the direct call is then
+/// modified to have call traits for a non-dispatch variant.
+static Expr *replaceWithNewTraitsOrDirectCall(const ASTContext &Context,
+                                              Expr *AssocExpr,
+                                              SemaOpenMP *SemaPtr,
+                                              bool NoContext) {
+  BinaryOperator *BinaryCopyOpr = nullptr;
+  bool IsBinaryOp = false;
+  Expr *PseudoObjExprOrCall = AssocExpr;
+  if (auto *BinOprExpr = dyn_cast<BinaryOperator>(AssocExpr)) {
+    IsBinaryOp = true;
+    BinaryCopyOpr = BinaryOperator::Create(
+        Context, BinOprExpr->getLHS(), BinOprExpr->getRHS(),
+        BinOprExpr->getOpcode(), BinOprExpr->getType(),
+        BinOprExpr->getValueKind(), BinOprExpr->getObjectKind(),
+        BinOprExpr->getOperatorLoc(), FPOptionsOverride());
+    PseudoObjExprOrCall = BinaryCopyOpr->getRHS();
+  }
+
+  Expr *CallWithoutInvariants = PseudoObjExprOrCall;
+  // Change PseudoObjectExpr to a direct call
+  if (auto *PseudoObjExpr = dyn_cast<PseudoObjectExpr>(PseudoObjExprOrCall))
+    CallWithoutInvariants = *((PseudoObjExpr->semantics_begin()) - 1);
+
+  Expr *FinalCall = CallWithoutInvariants; // For noinvariants clause
+  if (NoContext) {
+    // example to explain the changes done for "nocontext" clause:
+    //
+    // #pragma omp declare variant(foo_variant_dispatch)
+    // 		 		 match(construct = {dispatch})
+    // #pragma omp declare variant(foo_variant_allCond)
+    // 		 		match(user = {condition(1)})
+    // ...
+    //     #pragma omp dispatch nocontext(cond_true)
+    //         foo(i, j); // with traits: CodeGen call to
+    //         foo_variant_dispatch(i,j)
+    // dispatch construct is changed to:
+    // if (cond_true) {
+    //    foo(i,j) // with traits: CodeGen call to foo_variant_allCond(i,j)
+    // } else {
+    //   #pragma omp dispatch
+    //   foo(i,j)  // with traits: CodeGen call to foo_variant_dispatch(i,j)
+    // }
```

Comment:
Why you cannot do this directly in codegen?

Reply:
I am sorry for the detailed explanation below:

For a code like 
	```
#pragma omp dispatch nocontext(cond) depend(in:i)
	 foo(i,j)
```

`foo(i,j)` is present in PseudoObjectExpr, containing call to `foo(i,j)` and `foo_variant_dispatch(i,j)`.
The word `traits `used below is in relation to PseudoObjectExpr.

The Sema transformation can be seen as:
	```
#pragma omp taskwait depend(in:i)
	foo(i,j) --> with OMPDispatchNoContextAttr
```

In the transformation in Sema, instead of adding OMPDispatchNoContextAttr we change it to:
	```
if(cond) {
	   foo(i,j) --> with traits containing call to foo_variant_allCond(i,j)
	} else {
	   #pragma omp dispatch
	   foo(i,j) --> without any changes i.e. with traits containing 
                                              call to foo_variant_dispatch(i,j)
	}

```
in the then part, to change the function call from 
	```
         foo(i,j) --> with traits containing call to foo_variant_dispatch(i,j)
     to foo(i,j) --> with traits containing call to foo_variant_allCond(i,j)
```
  1) With,` foo(i,j)` (i.e. direct call without traits) 
  2) call `ActOnOpenMPCall()`. The `ActOnOpenMPCall()` will add the traits for call to `foo_variant_allCond(i,j)`.
  3) In the CodeGen, the call to `foo_variant_allCond(i,j) ` is emit-ed without any changes.

To do the same steps in CodeGen:
1) For every function call I need to check if it has `OMPDispatchNoContextAttr`.
2) Create a then part,
	i) With foo(i,j) (i.e. direct call) Either call `ActOnOpenMPCall()` to get
call to `foo_variant_dispatch(i,j) ` or replicate the steps from 7439 to 7486.	
	2) Create if-else statement as mentioned above for Sema, without `#pragma omp dispatch` in else-part.
        3) Emit the if-else statement.

Doing it in Sema is easier because functions, like `ActOnOpenMPCall()`, generating of PseudoObjectExpr are present in Sema itself.

Reply:
Sema should generate only required statements/expressions. If something can be emitted in codegen, better to emit in codegen, to avoid extra stuff in AST